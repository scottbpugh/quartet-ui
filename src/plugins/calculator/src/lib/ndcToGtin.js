// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

/**
 * Generates a check digit based on inbound character
 * strings for a given SSCC-18 or GTIN-14 barcode.
 *
 * @param barcode The first 13 characters or the
 * first 17 characters of an SSCC-18 or GTIN-14.
 */
export function mod10CheckDigit(barcode) {
  var totalOdd = 0;
  var totalEven = 0;
  var total = 0;
  // get the odd numbers
  for (var i = 0; i < barcode.length; i += 2) {
    var char = barcode.substr(i, 1);
    totalOdd += +char * 3;
  }
  // get the even numbers
  for (var i = 1; i < barcode.length; i += 2) {
    var char = barcode.substr(i, 1);
    totalEven += +char * 1;
  }
  total = totalOdd + totalEven;
  // see if this is a multiple of 10
  var near = Math.ceil(total / 10) * 10;
  console.debug(near);
  return Math.ceil(total / 10) * 10 - total;
}

/**
 * Converts an FDA NDC value into a valid GTIN14.
 * @param ndc An NDC with or without dashes- should always be 12 digits
 * with dashes or 10 digits without.
 */
export function NDCtoGTIN14(ndc, indicator) {
  // strip out the dashes and any white space
  var gtin = ndc.replace(/-/g, "").trim();
  gtin = indicator + "03" + gtin;
  var checkDigit = mod10CheckDigit(gtin);
  return "" + gtin + checkDigit;
}
//document.body.innerHTML = NDCtoGTIN14("99999-988-99", "1");
