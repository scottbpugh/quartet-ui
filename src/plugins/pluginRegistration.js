// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
import React from "react";
import {Route} from "react-router";
import {injectIntl} from "react-intl";

/**
 * PluginRegistry - Holds references to various plugin objects
 * such as routes, reducers, components, and localization messages.
 * Also emits an event when a reducer is added.
 */
class PluginRegistry {
  constructor() {
    this._emitChange = null;

    /*
    Holds references to reducers that come from plugins.
    */
    this._reducers = {};

    /*
    initialData isn't currently used due to a bug. All plugin state start as {}.
    Code your components to handle an empty object, or fix the combine function in the main store.js
    */
    this._initialData = {};

    /*
    Holds all the localization messages for plugins, when they are enabled.
    */
    this._messages = {"en-US": {plugins: {}}, "fr-FR": {plugins: {}}};

    /* Holds all the components from plugins that can be injected
    into the core components (buttons, context menus, ...)
    Use an event dispatch to tie them to a component (navtree, ...)
    For the sake of simplicity we don't have an object for each type
    of component. plugin name and component name are concatenated together for ease of access.
    */
    this._registeredComponents = {};

    /*
    We keep an entry previously registered and remove into _unregisteredComponents.
    This is to remove it from the parent elements where it has been added (e.g.: NavTreeItems, ButtonControls, etc, ...)
    */
    this._unregisteredComponents = {};

    /*
    Holds all the plugin routes and can replace an existing route.
    Please be considerate of others and use your plugin name in your path,
    e.g.: <Route path="/number-range/pools:serverID component={PoolList}/>
    */
    this._registeredRoutes = {};
  }

  setMessages(messages) {
    for (let language in messages) {
      if (language in this._messages) {
        this._messages[language].plugins = {
          ...this._messages[language].plugins,
          ...messages[language].plugins
        };
      } else {
        this._messages[language] = {plugins: {...messages.plugins}};
      }
    }
  }

  registerComponent = (pluginName, component, injectAction) => {
    if (!pluginName || !component) {
      throw new Error(
        "You must enter a valid pluginName string as well as a reference to a component to register a component."
      );
    }
    this._registeredComponents[
      `plugin_${pluginName}_${component.PLUGIN_COMPONENT_NAME}`
    ] = {
      pluginName: pluginName,
      component: component,
      action: injectAction
    };
    try {
      // attempt to remove components from unregistered, if the plugin was enabled/disabled prior.
      delete this._unregisteredComponents[
        `plugin_${pluginName}_${component.PLUGIN_COMPONENT_NAME}`
      ];
    } catch (e) {
      // ignore if this is the first time.
    }
  };

  unregisterComponent = (pluginName, component, injectAction) => {
    if (!pluginName || !component) {
      throw new Error(
        "You must enter a valid pluginName string as well as a reference to a component to register a component."
      );
    }
    let fullPluginComponentName = `plugin_${pluginName}_${
      component.PLUGIN_COMPONENT_NAME
    }`;
    this._unregisteredComponents[fullPluginComponentName] = {
      pluginName: pluginName,
      component: component,
      action: injectAction
    };

    try {
      delete this._registeredComponents[
        `plugin_${pluginName}_${component.PLUGIN_COMPONENT_NAME}`
      ];
    } catch (e) {
      //don't bother if it doesn't exist.
    }
  };

  getUnregisteredComponents = () => {
    return {...this._unregisteredComponents};
  };

  getRegisteredComponents = () => {
    return {...this._registeredComponents};
  };

  getRegisteredComponent = fullPluginComponentName => {
    try {
      return injectIntl(
        this._registeredComponents[fullPluginComponentName].component
      );
    } catch (e) {}
  };

  registerRoutes = (pluginName, routes) => {
    this._registeredRoutes[pluginName] = routes;
  };

  unregisterRoutes = pluginName => {
    delete this._registeredRoutes[pluginName];
  };

  getRegisteredRoutes = () => {
    return {...this._registeredRoutes};
  };

  getArrayRoutes = () => {
    let routes = [];
    for (let plugin in this._registeredRoutes) {
      routes = routes.concat(this._registeredRoutes[plugin]);
    }
    return routes;
  };

  getMessages() {
    return {...this._messages};
  }

  setInitialData(name, dataObject) {
    this._initialData[name] = dataObject;
  }

  getInitialData() {
    return {...this._initialData};
  }

  getReducers() {
    return {...this._reducers};
  }

  registerReducer(pluginName, name, reducer, initialData) {
    this.setInitialData(name, initialData);
    this._reducers = {...this._reducers, [name]: reducer};
    if (this._emitChange) {
      this._emitChange(this.getReducers());
    }
  }

  setChangeListener(listener) {
    this._emitChange = listener;
  }
}

// export only a singleton. Entire app and all plugins share one PluginRegistry object.
export const pluginRegistry = new PluginRegistry();
