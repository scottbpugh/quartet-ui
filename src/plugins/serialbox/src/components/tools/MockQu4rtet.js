// Copyright (c) 2018 Serial Lab
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
// Fake local storage.
const qu4rtetRequire = require("qu4rtet-ui/lib/require.js").default;

class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value;
  }

  removeItem(key) {
    delete this.store[key];
  }
}
window = global;
export const localStorage = new LocalStorageMock();

// add fake local storage for tests.
window.localStorage = localStorage;

// add fake require for ipcRenderer.
window.require = module => {
  if (module === "path") {
    return {
      join: () => {
        return "";
      }
    };
  } else if (module === "electron") {
    return {
      ipcRenderer: {
        on: () => {},
        send: () => {}
      },
      remote: {
        require: path => {
          if (path === "./main-process/plugin-manager.js") {
            return {
              install: () => {
                return {done: "ok"};
              }
            };
          }
          return {};
        },
        app: {
          getVersion: () => {
            return "test-version";
          },
          getPath: () => {
            return "somepath";
          }
        }
      }
    };
  } else {
    try {
      return qu4rtetRequire(module);      
    } catch(e) {
      return null;
    }

  }
};

const {qu4rtet} = require("qu4rtet-ui/lib/qu4rtet");
const React = qu4rtetRequire("react");
const {Component} = React;
