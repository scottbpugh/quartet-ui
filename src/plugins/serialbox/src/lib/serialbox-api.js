// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
const {showMessage} = qu4rtet.require("./lib/message");
const {pluginRegistry} = qu4rtet.require("./plugins/pluginRegistration");

const PREFIX_PATH = "serialbox/";

/**
 * getPools - Description
 *
 * @param {object} server A serversetting object.
 *
 * @return {object} A JSON object.
 */
export const getPools = server => {
  return pluginRegistry
    .getServer(server.serverID)
    .fetchListAll("serialbox_pools_list", {related: "true"}, [])
    .then(pools => {
      // console.log(pools)
      return pools;
    })
    .catch(error => {
      showMessage({
        type: "danger",
        id: "plugins.numberRange.errorFetchPools",
        values: {serverName: server.serverSettingName}
      });
      return error;
    });
};


export const getReponseRules = (server, pool) => {
  return pluginRegistry
}

/**
 * getRegion - Description
 *
 * @param {object} server   A serversetting Object
 * @param {string} regionName A region machine name
 *
 * @return {object} A JSON response.
 */
export const getRegion = (server, regionName) => {
  return getRegionByName(
    server,
    regionName,
    "serialbox_sequential_region_detail_read"
  );
};

/**
 * getRegionByName - Similar to getRegion but no URL logic.
 *
 * @param {object} server Server setting object.
 * @param {string} url The name of the region.
 * @param {string} operationId the operationId for the endpoint.
 *
 * @return {object} A JSON object.
 */
export const getRegionByName = async (server, name, operationId) => {
  try {
    let res = await pluginRegistry
      .getServer(server.serverID)
      .fetchObject(operationId, {machine_name: name});
    return res;
  } catch (error) {
    showMessage({
      type: "danger",
      id: "plugins.numberRange.errorFetchRegion",
      values: {error: error}
    });
    return error;
  }
};

/**
 * getRegions - Description
 *
 * @param {object} server Server setting object.
 * @param {object} pool   Pool object
 *
 * @return {Promise} A promise
 */
export const getRegions = (server, pool) => {
  let promises = [];
  for (let name of pool.sequentialregion_set) {
    promises.push(
      getRegionByName(server, name, "serialbox_sequential_region_detail_read")
    );
  }
  if (pool.randomizedregion_set) {
    for (let name of pool.randomizedregion_set) {
      promises.push(
        getRegionByName(server, name, "serialbox_randomized_regions_read")
      );
    }
  }

  if (pool.listbasedregion_set) {
    for (let name of pool.listbasedregion_set) {
      promises.push(
        getRegionByName(server, name, "serialbox_list_based_region_detail_read")
      );
    }
  }
  return Promise.all(promises)
    .then(data => {
      return data;
    })
    .catch(error => {
      showMessage({
        type: "danger",
        id: "plugins.numberRange.errorFetchRegion",
        values: {error: error}
      });
      return error;
    });
};


const prepHeadersAuth = async (
  server,
  method = "GET",
  contentType = "application/json",
  accept = "application/json"
) => {
  const headers = new Headers();
  headers.append("Accept", accept);
  if (contentType) {
    headers.append("Content-Type", contentType);
  }
  headers.append(
    "Authorization",
    await pluginRegistry.getServer(server.serverID).getAuthorization()
  );
  console.log("headers", JSON.stringify(headers));
  return {
    method,
    headers,
    mode: "cors"
  };
};


/**
 * allocate - Allocates numbers to pool
 *
 * @param {object} server A server setting object
 * @param {object} pool   A pool object
 * @param {integer} value  A numeric value.
 *
 * @return {object} A JSON response.
 */
export const allocate = async (server, pool, value, exportType) => {

  try {
    if (exportType === "xml") {
      // Requesting XML as is.
      let headers = await prepHeadersAuth(server, "GET", "application/xml", "application/xml");
      let response = await fetch(`${server.url}serialbox/allocate/${pool.machine_name}/${value}/`, {...headers});
      let raw = await response.text();
      if (!response.ok) {
	throw new Error(raw.substring(0,250))
      }
      return raw;
    } else {
      return await pluginRegistry
	.getServer(server.serverID)
	.fetchObject("serialbox_allocate_read", {
          pool: pool.machine_name,
          size: value
	});
    }
  } catch (error) {
    if (error.response && error.response.obj && error.response.obj.detail) {
      showMessage({
        type: "danger",
        msg: error.response.obj.detail
      });
    } else {
      showMessage({
        type: "danger",
        id: "plugins.numberRange.errorAllocating",
        values: {error: error, poolName: pool.readable_name}
      });
    }
    return error;
  }
};

export const postAddRegion = async (
  server, 
  postValues, 
  edit = false
  ) => {
  let method = "POST";
  let endpoint = "sequential-region-create";
  if (edit) {
    method = "PUT";
    endpoint = `sequential-region-modify/${postValues.machine_name}`;
  }
  let req = {method: method};
  req.body = JSON.stringify(postValues);
  return await pluginRegistry
    .getServer(server.serverID)
    .fetchWithHeaders(`${server.url}${PREFIX_PATH}${endpoint}/`, req)
    .then(resp => {
      return resp;
    })
    .catch(error => {
      showMessage({
        type: "error",
        id: "plugins.numberRange.errorVanilla",
        values: {error: error}
      });
    });
};
export const postAddListBasedRegion = async (
  server, 
  postValues, 
  edit = false
  ) => {
  let method = "POST";
  let endpoint = "list-based-region-create";
  if (edit) {
    method = "PUT";
    endpoint = `list-based-region-modify/${postValues.machine_name}`;
  }
  let req = {method: method};
  req.body = JSON.stringify(postValues);
  return await pluginRegistry
    .getServer(server.serverID)
    .fetchWithHeaders(`${server.url}${PREFIX_PATH}${endpoint}/`, req)
    .then(resp => {
      return resp;
    })
    .catch(error => {
      showMessage({
        type: "error",
        id: "plugins.numberRange.errorVanilla",
        values: {error: error}
      });
      throw error;
    });
};
export const postAddRandomizedRegion = async (
  server,
  postValues,
  edit = false
) => {
  let method = "POST";
  let endpoint = "randomized-regions";
  if (edit) {
    method = "PUT";
    endpoint = `randomized-regions/${postValues.machine_name}`;
  }
  let req = {method: method};
  req.body = JSON.stringify(postValues);
  return await pluginRegistry
    .getServer(server.serverID)
    .fetchWithHeaders(`${server.url}${PREFIX_PATH}${endpoint}/`, req)
    .then(resp => {
      return resp;
    })
    .catch(error => {
      showMessage({
        type: "error",
        id: "plugins.numberRange.errorVanilla",
        values: {error: error}
      });
      throw error;
    });
};
const getRegionType = region => {
  if (region.state) {
    return "serial";
  }
  if (!region.state && region.max) {
    return "randomized";
  }
  if (region.processing_class_path !== undefined) {
    return "list-based";
  }
  return null;
};

export const deleteRegion = async (server, region) => {
  let method = "DELETE";
  let endpoint = "";
  let regionType = getRegionType(region);
  if (regionType === "serial") {
    // sequential
    endpoint = `sequential-region-modify/${region.machine_name}`;
  } else if (regionType === "randomized") {
    // randomized
    endpoint = `randomized-regions/${region.machine_name}`;
  } else if (regionType === "list-based") {
    endpoint = `list-based-region-modify/${region.machine_name}`;
  }
  let req = {method: method};
  return await pluginRegistry
    .getServer(server.serverID)
    .fetchWithHeaders(`${server.url}${PREFIX_PATH}${endpoint}/`, req)
    .then(resp => {
      return resp;
    })
    .catch(error => {
      showMessage({
        type: "error",
        id: "plugins.numberRange.errorVanilla",
        values: {error: error}
      });
      throw error;
    });
};

export const deletePool = async (server, pool) => {
  let method = "DELETE";
  let endpoint = `pool-modify/${pool.machine_name}`;
  let req = {method: method};

  return await pluginRegistry
    .getServer(server.serverID)
    .fetchWithHeaders(`${server.url}${PREFIX_PATH}${endpoint}/`, req)
    .then(resp => {
      return resp;
    })
    .catch(error => {
      showMessage({
        type: "error",
        id: "plugins.numberRange.errorVanilla",
        values: {error: error}
      });
      throw error;
    });
};

export const postAddPool = async (server, postValues, edit = false) => {
  let endpoint = "pool-create";
  let method = "POST";
  if (edit) {
    method = "PUT";
    endpoint = `pool-modify/${postValues.machine_name}`;

    try {
      // remove sets before submitting.
      delete postValues.sequentialregion_set;
      delete postValues.randomizedregion_set;
    } catch (e) {
      console.log(e);
    }
  }
  let req = {method: method};
  req.body = JSON.stringify(postValues);
  return await pluginRegistry
    .getServer(server.serverID)
    .fetchWithHeaders(`${server.url}${PREFIX_PATH}${endpoint}/`, req)
    .then(resp => {
      return resp;
    })
    .catch(error => {
      showMessage({
        type: "error",
        id: "plugins.numberRange.errorVanilla",
        values: {error: error}
      });
      throw error;
    });
};
