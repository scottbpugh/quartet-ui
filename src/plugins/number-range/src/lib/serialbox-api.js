// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
import base64 from "base-64";
import {showMessage} from "lib/message";

const PREFIX_PATH = "serialbox/";

/**
 * prepHeaders - Prepares the headers to be sent.
 *
 * @param {object} server A server setting object
 *
 * @return {object} A request init object with headers.
 */
const prepHeaders = (server, method = "GET") => {
  let username = server.username;
  let password = server.password;
  let headers = new Headers();
  headers.append("Accept", "application/json");
  headers.append("Content-Type", "application/json");
  headers.append(
    "Authorization",
    "Basic " + base64.encode(username + ":" + password)
  );
  return {
    method: method,
    headers: headers,
    mode: "cors"
  };
};

/**
 * getPools - Description
 *
 * @param {object} server A serversetting object.
 *
 * @return {object} A JSON object.
 */
export const getPools = server => {
  const url = `${server.url}${PREFIX_PATH}pools/?related=true`;
  return fetch(url, prepHeaders(server))
    .then(resp => {
      return resp.json();
    })
    .then(data => {
      return data;
    })
    .catch(error => {
      showMessage({
        type: "danger",
        id: "plugins.numberRange.errorFetchPools",
        values: {serverName: server.serverSettingName}
      });
      return error;
    });
};

export const getPoolDetail = (server, pool) => {
  const url = `${server.url}${PREFIX_PATH}pool-detail/${
    pool.machine_name
  }/?related=true`;
  return fetch(url, prepHeaders(server))
    .then(resp => {
      return resp.json();
    })
    .then(data => {
      return data;
    })
    .catch(error => {
      showMessage({
        type: "danger",
        id: "plugins.numberRange.errorFetchPool",
        values: {poolName: pool.readable_name, error: error}
      });
      return error;
    });
};

/**
 * getRegion - Description
 *
 * @param {object} server   A serversetting Object
 * @param {string} regionName A region machine name
 *
 * @return {object} A JSON response.
 */
export const getRegion = (server, regionName) => {
  return getRegionByURL(
    server,
    `${server.url}${PREFIX_PATH}sequential-region-detail/${regionName}/`
  );
};

/**
 * getRegionByURL - Similar to getRegion but no URL logic.
 *
 * @param {object} server Server setting object.
 * @param {string} url A full URL to the API endpoint.
 *
 * @return {object} A JSON object.
 */
export const getRegionByURL = (server, url) => {
  return fetch(url, prepHeaders(server))
    .then(resp => {
      return resp.json();
    })
    .then(data => {
      return data;
    })
    .catch(error => {
      showMessage({
        type: "danger",
        id: "plugins.numberRange.errorFetchRegion",
        values: {error: error}
      });
      return error;
    });
};

/**
 * getRegions - Description
 *
 * @param {object} server Server setting object.
 * @param {object} pool   Pool object
 *
 * @return {Promise} A promise
 */
export const getRegions = (server, pool) => {
  let promises = [];
  for (let url of pool.sequentialregion_set) {
    promises.push(getRegionByURL(server, url));
  }
  if (pool.randomizedregion_set) {
    for (let url of pool.randomizedregion_set) {
      promises.push(getRegionByURL(server, url));
    }
  }
  return Promise.all(promises)
    .then(data => {
      return data;
    })
    .catch(error => {
      showMessage({
        type: "danger",
        id: "plugins.numberRange.errorFetchRegion",
        values: {error: error}
      });
      return error;
    });
};

/**
 * allocate - Allocates numbers to pool
 *
 * @param {object} server A server setting object
 * @param {object} pool   A pool object
 * @param {integer} value  A numeric value.
 *
 * @return {object} A JSON response.
 */
export const allocate = (server, pool, value) => {
  return fetch(
    `${server.url}${PREFIX_PATH}allocate/${pool.machine_name}/${value}/`,
    prepHeaders(server)
  )
    .then(resp => {
      return resp.json();
    })
    .then(data => {
      return data;
    })
    .catch(error => {
      showMessage({
        type: "danger",
        id: "plugins.numberRange.errorAllocating",
        values: {error: error, poolName: pool.readable_name}
      });
      throw error;
    });
};

export const getRegionFormStructure = server => {
  return fetch(
    `${server.url}${PREFIX_PATH}sequential-region-create/`,
    prepHeaders(server, "OPTIONS")
  )
    .then(resp => {
      return resp.json();
    })
    .then(data => {
      return data;
    })
    .catch(error => {
      showMessage({
        type: "danger",
        id: "plugins.numberRange.errorAllocating",
        values: {error: error}
      });
      throw error;
    });
};

export const getRandomizedRegionFormStructure = server => {
  return fetch(
    `${server.url}${PREFIX_PATH}randomized-regions/`,
    prepHeaders(server, "OPTIONS")
  )
    .then(resp => {
      return resp.json();
    })
    .then(data => {
      return data;
    })
    .catch(error => {
      showMessage({
        type: "danger",
        id: "plugins.numberRange.errorFormFetch",
        values: {error: error, serverName: server.serverSettingName}
      });
      throw error;
    });
};

export const getPoolFormStructure = server => {
  return fetch(
    `${server.url}${PREFIX_PATH}pool-create/`,
    prepHeaders(server, "OPTIONS")
  )
    .then(resp => {
      return resp.json();
    })
    .then(data => {
      return data;
    })
    .catch(error => {
      showMessage({
        type: "danger",
        id: "plugins.numberRange.errorFormFetch",
        values: {error: error, serverName: server.serverSettingName}
      });
      throw error;
    });
};

export const postAddRegion = (server, postValues, edit = false) => {
  let method = "POST";
  let endpoint = "sequential-region-create";
  if (edit) {
    method = "PUT";
    endpoint = `sequential-region-modify/${postValues.machine_name}`;
  }
  let headers = prepHeaders(server, method);
  headers.body = JSON.stringify(postValues);
  return fetch(`${server.url}${PREFIX_PATH}${endpoint}/`, headers)
    .then(resp => {
      return resp;
    })
    .catch(error => {
      showMessage({
        type: "error",
        id: "plugins.numberRange.errorVanilla",
        values: {error: error}
      });
    });
};

export const postAddRandomizedRegion = (server, postValues, edit = false) => {
  let method = "POST";
  let endpoint = "randomized-regions";
  if (edit) {
    method = "PUT";
    endpoint = `randomized-regions/${postValues.machine_name}`;
  }
  let headers = prepHeaders(server, method);
  headers.body = JSON.stringify(postValues);
  return fetch(`${server.url}${PREFIX_PATH}${endpoint}/`, headers)
    .then(resp => {
      return resp;
    })
    .catch(error => {
      showMessage({
        type: "error",
        id: "plugins.numberRange.errorVanilla",
        values: {error: error}
      });
      throw error;
    });
};

export const deleteRegion = (server, region) => {
  let method = "DELETE";
  let endpoint = "";
  if (region.state) {
    // sequential
    endpoint = `sequential-region-modify/${region.machine_name}`;
  } else if (region.remaining || region.remaining === 0) {
    // randomized
    endpoint = `randomized-regions/${region.machine_name}`;
  }
  let headers = prepHeaders(server, method);
  //headers.body = JSON.stringify(postValues);
  return fetch(`${server.url}${PREFIX_PATH}${endpoint}/`, headers)
    .then(resp => {
      return resp;
    })
    .catch(error => {
      showMessage({
        type: "error",
        id: "plugins.numberRange.errorVanilla",
        values: {error: error}
      });
      throw error;
    });
};

export const deletePool = (server, pool) => {
  let method = "DELETE";
  let endpoint = `pool-modify/${pool.machine_name}`;
  let headers = prepHeaders(server, method);
  //headers.body = JSON.stringify(postValues);
  return fetch(`${server.url}${PREFIX_PATH}${endpoint}/`, headers)
    .then(resp => {
      return resp;
    })
    .catch(error => {
      showMessage({
        type: "error",
        id: "plugins.numberRange.errorVanilla",
        values: {error: error}
      });
      throw error;
    });
};

export const postAddPool = (server, postValues, edit = false) => {
  let endpoint = "pool-create";
  let method = "POST";
  if (edit) {
    method = "PUT";
    endpoint = `pool-modify/${postValues.machine_name}`;
  }
  let headers = prepHeaders(server, method);
  headers.body = JSON.stringify(postValues);
  return fetch(`${server.url}${PREFIX_PATH}${endpoint}/`, headers)
    .then(resp => {
      return resp;
    })
    .catch(error => {
      showMessage({
        type: "error",
        id: "plugins.numberRange.errorVanilla",
        values: {error: error}
      });
      throw error;
    });
};
