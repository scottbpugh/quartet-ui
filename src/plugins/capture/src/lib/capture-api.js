// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import {showMessage} from "lib/message";
import base64 from "base-64";
import {pluginRegistry} from "plugins/pluginRegistration";

export const fileUpload = (server, rule, fileObject) => {
  let data = new FormData();
  let headers = new Headers();
  headers.append(
    "Authorization",
    "Basic " + base64.encode(server.username + ":" + server.password)
  );
  headers.append("Accept", "application/json");
  data.append("file", fileObject);
  fetch(`${server.url}capture/quartet-capture/?rule=${rule.name}`, {
    method: "POST",
    headers: headers,
    body: data
  })
    .then(resp => {
      resp
        .json()
        .then(data => {
          if (!resp.ok) {
            showMessage({type: "error", msg: data.detail});
          } else {
            showMessage({type: "success", msg: data});
          }
        })
        .catch(e => {
          showMessage({
            type: "error",
            id: "plugins.capture.processFileError"
          });
        });
    })
    .catch(resp => {
      showMessage({
        type: "error",
        id: "plugins.capture.processFileError"
      });
    });
};

export const getRuleParamsByRule = (server, ruleID) => {
  // this is a temporary workaround to get all the rule params for a given ruleID.
  // Eventually we should switch to sending the id of the rule and get back a subset.
  return pluginRegistry
    .getServer(server.serverID)
    .getClient()
    .then(client => {
      return client.apis.capture
        .capture_rule_parameters_list()
        .then(result => {
          if (result.ok) {
            // filter result.
            return result.body.filter(param => {
              if (param.rule === ruleID) {
                return true;
              }
              return false;
            });
          } else {
            showMessage({
              type: "error",
              id: "plugins.capture.errorFetchRuleParams"
            });
            return [];
          }
        })
        .catch(e => {
          showMessage({
            type: "error",
            id: "plugins.capture.errorFetchRuleParams"
          });
        });
    });
};
