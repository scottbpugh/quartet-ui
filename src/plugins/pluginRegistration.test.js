// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
import "tools/mockStore"; // mock ipcRenderer, localStorage, ...

import {Server} from "lib/servers";
import {pluginRegistry} from "./pluginRegistration";

it("adds a server when registering a server", () => {
  pluginRegistry.registerServer(new Server({serverID: "random-id"}));
  expect(pluginRegistry.getServer("random-id")).toMatchSnapshot();
  expect(Object.keys(pluginRegistry._servers).length).toEqual(1);
});

it("removes a server when triggering removeServer", () => {
  pluginRegistry.registerServer(new Server({serverID: "random-id"}));
  expect(pluginRegistry.getServer("random-id")).toMatchSnapshot();
  // making sure the server is properly registered here one more time.
  expect(Object.keys(pluginRegistry._servers).length).toEqual(1);
  pluginRegistry.removeServer(pluginRegistry.getServer("random-id"));
  expect(Object.keys(pluginRegistry._servers).length).toEqual(0);
});

it("can retrieve a server based on server object or serverID", () => {
  const s = new Server({serverID: "random-id"});
  pluginRegistry.registerServer(s);
  expect(pluginRegistry.getServer(s)).toMatchSnapshot();
});
