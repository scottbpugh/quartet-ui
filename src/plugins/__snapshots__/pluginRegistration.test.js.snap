// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adds a server when registering a server 1`] = `
Object {
  "authType": "basic",
  "hostname": undefined,
  "path": "",
  "port": 80,
  "protocol": "http",
  "serverID": "random-id",
  "serverSettingName": undefined,
  "ssl": undefined,
  "tokenType": "Token",
  "url": "http://undefined:80/",
  "username": undefined,
}
`;

exports[`can retrieve a server based on server object or serverID 1`] = `
Object {
  "authType": "basic",
  "hostname": undefined,
  "path": "",
  "port": 80,
  "protocol": "http",
  "serverID": "random-id",
  "serverSettingName": undefined,
  "ssl": undefined,
  "tokenType": "Token",
  "url": "http://undefined:80/",
  "username": undefined,
}
`;

exports[`removes a server when triggering removeServer 1`] = `
Object {
  "authType": "basic",
  "hostname": undefined,
  "path": "",
  "port": 80,
  "protocol": "http",
  "serverID": "random-id",
  "serverSettingName": undefined,
  "ssl": undefined,
  "tokenType": "Token",
  "url": "http://undefined:80/",
  "username": undefined,
}
`;
